agent/eventhandler/attachment_handler.go

```diff
@@ -0,0 +1,134 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package eventhandler
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+
+	"github.com/aws/amazon-ecs-agent/agent/api"
+	"github.com/aws/amazon-ecs-agent/agent/statechange"
+	"github.com/aws/amazon-ecs-agent/agent/statemanager"
+	"github.com/aws/amazon-ecs-agent/agent/utils/retry"
+	"github.com/cihub/seelog"
+)
+
+// AttachmentHandler is a handler that is responsible for submitting attachment state change events
+// to backend
+type AttachmentHandler struct {
+	// stateSaver is a statemanager which may be used to save any
+	// changes to an attachment's SentStatus
+	stateSaver statemanager.Saver
+
+	// attachmentLock is a map from attachment ARN to a lock.
+	// we need to have a lock for each attachment so that the attached
+	// status of an attachment won't be sent multiple times
+	attachmentLock map[string]*sync.Mutex
+
+	// lock is used to safely access the attachmentLock map
+	lock sync.Mutex
+
+	client api.ECSClient
+	ctx    context.Context
+}
+
+// NewAttachmentHandler returns a new AttachmentHandler object
+func NewAttachmentHandler(ctx context.Context,
+	stateSaver statemanager.Saver,
+	client api.ECSClient) *AttachmentHandler {
+	return &AttachmentHandler{
+		ctx:            ctx,
+		stateSaver:     stateSaver,
+		client:         client,
+		attachmentLock: make(map[string]*sync.Mutex),
+	}
+}
+
+// AddStateChangeEvent adds a state change event to AttachmentHandler for it to handle
+func (handler *AttachmentHandler) AddStateChangeEvent(change statechange.Event) error {
+	if change.GetEventType() != statechange.AttachmentEvent {
+		return errors.New(fmt.Sprintf("eventhandler: attachment handler received unrecognized event type: %d", change.GetEventType()))
+	}
+	event, ok := change.(api.AttachmentStateChange)
+	if !ok {
+		return errors.New("eventhandler: unable to get attachment event from state change event")
+	}
+
+	if event.Attachment == nil {
+		return fmt.Errorf("eventhandler: received malformed attachment state change event: %v", event)
+	}
+
+	go handler.submitAttachmentEvent(&event)
+	return nil
+}
+
+// submitAttachmentEvent submits an attachment state change to backend
+func (handler *AttachmentHandler) submitAttachmentEvent(attachmentChange *api.AttachmentStateChange) {
+	attachmentARN := attachmentChange.Attachment.AttachmentARN
+
+	var atmLock *sync.Mutex // the lock for the attachment
```

Comment:
This seems like a recipe for a future deadlock. I'd strongly suggest evaluating alternatives. One of which could be that instead of a single handler handling all attachments, you have a handler per attachment so that you don't have acquire 2 levels of locks. We've have multiple deadlocks in the agent in the past because of this pattern.

Reply:
refactored the attachment handler. now it has a map of attachment handlers each handling a certain attachment and has its own lock. there are still 2 level of locks but they won't be acquired together or affecting each other.